### DISCUSSION TITLE
Custom damage

### DISCUSSION THREAD

**Discussion by wehrad** [idaholab/moose#20473] [Q&A Getting Started] [answered]
Hi everyone,

as a new MOOSE user, I am trying to implement, using MASTODON, my own equation for a custom damage as a function of the von Mises stress and some constants and defined as a first order variable across my domain. I found many examples for a damage as a function of (x, y, t) and its influence on stresses but couldn't find (most likely missed) an example that would help me in my application.

I tried to have damage as a new material variable and access the von Mises stress to compute it. However, I run into an error when I want to multiply the stress to a value because it is a tensor... While I think it should, in theory, simply be [a scalar](https://mooseframework.inl.gov/source/utils/RankTwoScalarTools.html)?

Files needed to replicate this error are available [here](https://github.com/AdrienWehrle/moose/tree/next/examples/ex22_customdamage).
```
error: cannot convert 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeTensor<double> >::type' {aka 'libMesh::TypeTensor<double>'} to 'libMesh::Real' {aka 'double'} in initialization
   49 |   Real Xi = _alpha * _von_mises;
      |             ~~~~~~~^~~~~~~~~~~~
      |                    |
      |                    libMesh::boostcopy::enable_if_c<true, libMesh::TypeTensor<double> >::type {aka libMesh::TypeTensor<double>}
```

I think I am probably already wrong in my strategy. Damage should probably be an auxiliary variable and not a material one. I took that decision as I would like to have the material viscosity to depend on the damage in the case of viscous deformation in the future. What do you think?

Any help and/or advice would be much appreciated! Thank you a lot!

**Comment by bwspenc** [comment]
I believe that in MOOSE/tensor_mechanics we only have the damage base class as well as a couple of simple models that are really only useful for testing purposes. It sounds like those are the ones that you found. I would suggest looking at the Mazars damage model for concrete, which is implemented in the open-source BlackBear code:
https://github.com/idaholab/blackbear/blob/devel/src/materials/MazarsDamage.C
It's a stress-dependent scalar damage model. I believe that you can optionally link BlackBear in with MASTODON if that model happens to suit your needs.

**Reply by wehrad** [reply]
Thank you a lot @bwspenc, this is very useful and exactly what I needed (a basis to start working with)!

**Comment by bwspenc** [comment, marked as answer]
I guess I didn't answer your question about why you got the compile error. You are coupling to a variable that provides a scalar, but to access the value of that at your quadrature point, you'd need to index into it like this: `_von_mises[_qp]`.

However, that isn't really the best way to get at the von Mises stress, because you're extracting that from the stress tensors at the quadrature points, and then converting that to a field variable, so you're losing information. It would be better to extract it directly as a scalar material property at your quadrature point, and then use that material property containing the von Mises stress in your damage model. The `RankTwoInvariant` material model in the tensor_mechanics module can give you the von Mises stress as a material property, and then you can grab that in your damage model using `getMaterialProperty()`.

**Reply by wehrad** [reply]
> However, that isn't really the best way to get at the von Mises stress, because you're extracting that from the stress tensors at the quadrature points, and then converting that to a field variable, so you're losing information. It would be better to extract it directly as a scalar material property at your quadrature point, and then use that material property containing the von Mises stress in your damage model. The RankTwoInvariant material model in the tensor_mechanics module can give you the von Mises stress as a material property, and then you can grab that in your damage model using getMaterialProperty(). 

Thank you a lot again, I haven't thought about using `RankTwoInvariant`, that indeed sounds like a much better and cleaner idea. Thank you a lot!

**Comment by wehrad** [comment]
Based on @bwspenc's inputs I worked on my own custom damage material starting from [BlackBear's `MazarsDamage`](https://github.com/idaholab/blackbear/blob/devel/src/materials/MazarsDamage.C). 

After some debugging to (try to) get everything right, I finally got my application to run. Unfortunately now I am running into a logic error: 
my damage variable is successfully initialized but somehow `updateQpDamageIndex` is never called. I tried to better understand what was happening through silly checks/prints in [`ScalarDamageBase.C`](https://github.com/idaholab/moose/blob/next/modules/tensor_mechanics/src/materials/ScalarDamageBase.C) but couldn't find what I'm doing wrong here... My damage update looks right and compiles but is apparently simply not called... Or I'm missing some important logs...

The custom damage object and my input file can be found [here](https://github.com/AdrienWehrle/moose/tree/next/examples/ex22_customdamage).

Any help would be much appreciated, thank you a lot for your already precious help!

**Reply by bwspenc** [reply]
I think that's just because you need to comment out the block for the `ComputeFiniteStrainElasticStress` model and uncomment the block for the `ComputeDamageStress` model in your input file.
